instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/EvRerouteTest.kt

```diff
@@ -0,0 +1,217 @@
+                            "ev_connector_types" to "tesla",
+                            "energy_consumption_curve" to "0,300;20,160;80,140;120,180",
+                            "ev_charging_curve" to "40000,70000",
+                        )
+                    )
+                    .build(),
+                object : NavigationRouterCallback {
+                    override fun onRoutesReady(
+                        routes: List<NavigationRoute>,
+                        routerOrigin: RouterOrigin
+                    ) {
+                        mapboxNavigation.setNavigationRoutes(routes)
+                    }
+
+                    override fun onFailure(
+                        reasons: List<RouterFailure>,
+                        routeOptions: RouteOptions
+                    ) {
+                        logE("onFailure reasons=$reasons", "DEBUG")
+                    }
+
+                    override fun onCanceled(
+                        routeOptions: RouteOptions,
+                        routerOrigin: RouterOrigin
+                    ) {
+                        logE("onCanceled", "DEBUG")
+                    }
+                }
+            )
+        }
+
+        // wait for tracking to start
+        mapboxHistoryTestRule.stopRecordingOnCrash("no location tracking") {
+            Espresso.onIdle()
+        }
+        locationTrackingIdlingResource.unregister()
+
+        // push off route location and wait for the off route event
+        offRouteIdlingResource.register()
+        runOnMainSync {
+            mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+        }
+
+        mapboxHistoryTestRule.stopRecordingOnCrash("no off route") {
+            Espresso.onIdle()
+        }
+        offRouteIdlingResource.unregister()
+
+        // wait for tracking to start again
+        locationTrackingIdlingResource.register()
+        mapboxHistoryTestRule.stopRecordingOnCrash("no tracking") {
+            Espresso.onIdle()
+        }
+        locationTrackingIdlingResource.unregister()
+
+        val countDownLatch = CountDownLatch(1)
+        runOnMainSync {
+            mapboxNavigation.historyRecorder.stopRecording {
+                logE("history path=$it", "DEBUG")
+                countDownLatch.countDown()
+            }
+        }
+        countDownLatch.await()
+
+        // assert results
+        expectedStates.assert()
+
+        runOnMainSync {
+            val routeOptions = mapboxNavigation.getNavigationRoutes().first().routeOptions
+            val newWaypoints = routeOptions.coordinatesList()
+
+            check(newWaypoints.size == 2) {
+                "Expected 2 waypoints in the route after reroute but was ${newWaypoints.size}"
+            }
+            check(newWaypoints[1] == mockRoute.routeWaypoints.last())
+            with(routeOptions.unrecognizedJsonProperties!!) {
+                check(get("engine")!!.asString == "electric")
```

Comment:
Why not `assertEquals`? IMO `check` is more for the code to check some preconditions and throw an exception if we are not in a compatible state.